File: src/test/java/com/koddy/server/coffeechat/presentation/CancelCoffeeChatApiTest.java
//package com.koddy.server.coffeechat.presentation;
//
//import com.koddy.server.coffeechat.application.usecase.CancelCoffeeChatUseCase;
//import com.koddy.server.coffeechat.presentation.request.CancelCoffeeChatRequest;
//import com.koddy.server.common.ApiDocsTest;
//import com.koddy.server.member.domain.model.mentor.Mentor;
//import org.junit.jupiter.api.DisplayName;
//import org.junit.jupiter.api.Nested;
//import org.junit.jupiter.api.Test;
//import org.springframework.beans.factory.annotation.Autowired;
//
//import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
//import static com.koddy.server.common.utils.RestDocsSpecificationUtils.SnippetFactory.path;
//import static com.koddy.server.common.utils.RestDocsSpecificationUtils.createHttpSpecSnippets;
//import static com.koddy.server.common.utils.RestDocsSpecificationUtils.successDocsWithAccessToken;
//import static org.mockito.ArgumentMatchers.any;
//import static org.mockito.Mockito.doNothing;
//import static org.springframework.restdocs.request.RequestDocumentation.pathParameters;
//import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
//
//@DisplayName("CoffeeChat -> CancelCoffeeChatApi 테스트")
//class CancelCoffeeChatApiTest extends ApiDocsTest {
//    @Autowired
//    private CancelCoffeeChatUseCase cancelCoffeeChatUseCase;
//
//    private static final Long COFFEE_CHAT_ID = 1L;
//    private final Mentor mentor = MENTOR_1.toDomain().apply(1L);
//
//    @Nested
//    @DisplayName("신청/제안한 커피챗 취소 API [DELETE /api/coffeechats/cancel/{coffeeChatId}]")
//    class Cancel {
//        private static final String BASE_URL = "/api/coffeechats/cancel/{coffeeChatId}";
//        private final CancelCoffeeChatRequest request = new CancelCoffeeChatRequest("취소..");
//
//        @Test
//        @DisplayName("신청/제안한 커피챗을 취소한다")
//        void success() {
//            // given
//            applyToken(true, mentor);
//            doNothing()
//                    .when(cancelCoffeeChatUseCase)
//                    .invoke(any());
//
//            // when - then
//            successfulExecute(
//                    patchRequestWithAccessToken(new UrlWithVariables(BASE_URL, COFFEE_CHAT_ID), request),
//                    status().isNoContent(),
//                    successDocsWithAccessToken("CoffeeChatApi/LifeCycle/Cancel", createHttpSpecSnippets(
//                            pathParameters(
//                                    path("coffeeChatId", "커피챗 ID(PK)", true)
//                            )
//                    ))
//            );
//        }
//    }
//}

File: src/test/java/com/koddy/server/coffeechat/presentation/CancelCoffeeChatApiTest.kt
package com.koddy.server.coffeechat.presentation

import com.koddy.server.coffeechat.application.usecase.CancelCoffeeChatUseCase
import com.koddy.server.coffeechat.exception.CoffeeChatException
import com.koddy.server.coffeechat.exception.CoffeeChatExceptionCode
import com.koddy.server.coffeechat.presentation.request.CancelCoffeeChatRequest
import com.koddy.server.common.ApiDocsTestKt
import com.koddy.server.common.docs.DocumentField
import com.koddy.server.common.docs.NUMBER
import com.koddy.server.common.docs.STRING
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.justRun
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest

@WebMvcTest(CancelCoffeeChatApi::class)
@DisplayName("CoffeeChat -> CancelCoffeeChatApi 테스트")
internal class CancelCoffeeChatApiTest : ApiDocsTestKt() {
    @MockkBean
    private lateinit var cancelCoffeeChatUseCase: CancelCoffeeChatUseCase

    @Nested
    @DisplayName("신청/제안한 커피챗 취소 API [PATCH /api/coffeechats/cancel/{coffeeChatId}]")
    internal inner class Cancel {
        private val baseUrl = "/api/coffeechats/cancel/{coffeeChatId}"
        private val request = CancelCoffeeChatRequest(cancelReason = "취소..")

        private val pathParameters: Array<DocumentField> = arrayOf(
            "coffeeChatId" type NUMBER means "커피챗 ID(PK)",
        )
        private val requestFields: Array<DocumentField> = arrayOf(
            "cancelReason" type STRING means "취소 사유",
        )

        @Test
        fun `해당 커피챗이 취소될 수 없는 상태면 취소가 불가능하다`() {
            val exceptionCode = CoffeeChatExceptionCode.CANNOT_CANCEL_STATUS
            every { cancelCoffeeChatUseCase.invoke(any()) } throws CoffeeChatException(exceptionCode)

            patchRequest(baseUrl, arrayOf(1L)) {
                accessToken(common)
                bodyContent(request)
            }.andExpect {
                status { isConflict() }
                content { exception(exceptionCode) }
            }.andDo {
                makeFailureDocsWithAccessToken("CoffeeChatApi/LifeCycle/Cancel/Failure") {
                    pathParameters(*pathParameters)
                    requestFields(*requestFields)
                }
            }
        }

        @Test
        fun `커피챗을 취소한다`() {
            justRun { cancelCoffeeChatUseCase.invoke(any()) }

            patchRequest(baseUrl, arrayOf(1L)) {
                accessToken(common)
                bodyContent(request)
            }.andExpect {
                status { isNoContent() }
            }.andDo {
                makeSuccessDocsWithAccessToken("CoffeeChatApi/LifeCycle/Cancel/Success") {
                    pathParameters(*pathParameters)
                    requestFields(*requestFields)
                }
            }
        }
    }
}
